---
layout: post
title:  "HW4: Finding the same cell type in Eevee data"
author: Suki Ogihara
jhed: iogihar1
categories: [ HW EC1 ]
image: homework/hwEC1/hwEC1_iogihar1.gif
featured: false
---

<!-- Create a multi-panel data visualization that includes at minimum the following components:
A panel visualizing your one cluster of interest in reduced dimensional space (PCA, tSNE, etc)
A panel visualizing your one cluster of interest in physical space
A panel visualizing differentially expressed genes for your cluster of interest
A panel visualizing one of these genes in reduced dimensional space (PCA, tSNE, etc)
A panel visualizing one of these genes in space

Describe your figure briefly so we know what you are depicting (you no longer need to use precise data visualization terms as you have been doing). Write a description to convince me that your cluster interpretation is correct. Your description may reference papers and content that allowed you to interpret your cell cluster as a particular cell-type. You must provide attribution to external resources referenced. Links are fine; formatted references are not required. You must include the entire code you used to generate the figure so that it can be reproduced.  -->

### What's vizualized?
* A panel visualizing the cluster derived with kmeans in reduced dimensional space using linear vs non-linear methods (PCA and tSNE), as well as in the original physical space.  
By appliying animation between these three panels, you can observe the difference in how two different methods clusters differently in the reduced dimensional space. Interesting to note that in this particular case, the PCA method seems to cluster the data more clearly than tSNE.


```r 
library(ggplot2)
library(stats) # for kmeans
library(reshape2) # for melt
library(Rtsne) # for tSNE
library(gganimate) # for animation
set.seed(42)

# Change this to your local directory where the data is stored
dir <-"./genomic-data-visualization-2025/"
pikachu_file <- "data/pikachu.csv.gz"
data_P <- read.csv(paste0(dir, pikachu_file))

# Picachu data
gexp_P <- data_P[, 7:ncol(data_P)]
# log transform the gexp data
log_gexp_P <- log2(gexp_P+1)
pos_P <- data_P[, 5:6]
# PCA
pca_P <- prcomp(log_gexp_P, scale=TRUE)
# tSNE
# emb_P <- Rtsne(log_gexp_P, dims=3, pca = TRUE, perplexity=30, verbose=FALSE)
# rename colums of the tSNE data
colnames(emb_P$Y) <- c("tSNE1", "tSNE2", "tSNE3")
# kmeans clustering
# do kmeans clustering on the log transformed data
k=7
kmeans_orig_P <- kmeans(log_gexp_P, centers=k)
clusters_P <- as.factor(kmeans_orig_P$cluster)

# add the x-y coordinates to the data frame
# df_P <- cbind(data_P[, 1:6], log_gexp_P, pca_P$x[, 1:3], emb_P$Y, clusters_P)
# head(df_P)

df_1 <- data.frame(x = pos_P$aligned_x, y = pos_P$aligned_y, cluster = clusters_P)
df_2 <- data.frame(x = pca_P$x[,1], y = pca_P$x[,2], cluster = clusters_P)
df_3 <- data.frame(x = emb_P$Y[,1], y = emb_P$Y[,2], cluster = clusters_P)

df <- rbind(cbind(df_1, order=1), cbind(df_2, order = 2), cbind(df_3, order = 3))

p <- ggplot(df, aes(x = x, y = y, color = cluster)) + geom_point(size=.75) # + facet_wrap(~order) + theme_minimal()

anim <- p + transition_states(order) + view_follow() + enter_fade() + exit_fade() + labs(title = 'PCA vs tSNE') + theme_minimal()
gif <- animate(anim, height = 800, width = 800)
gif
```